### ISSUE TITLE
Help needed with custom Flipnote style pattern brushes

### ISSUE THREAD

**Issue by GoodbyeGhost** [opentoonz/opentoonz#4229]
## The Goal
I've been wanting to replicate the style of Flipnotes in my animation for a while now, but don't know how I would go about making custom brushes, if it's possible to do so with what I want. I'm aiming to replicate the 10 dither brushes found in this Krita package: https://sabs.like.sobs.moe/post/185842858590/hello-this-is-a-small-brush-pack-for-the-free-art. It's probably best if you use the pack to better understand what I'm wanting to make.

If making just the 10 dither brushes isn't too crazy of a thing, I wouldn't be apposed to the idea of recreating the entire package for Opentoonz either. I have tons of free time pitching in if needed, and really want this to happen, as I'm sure there are more people than myself who like this style and would welcome Opentoonz into their workflow if these brushes were available. 

I don't have first-hand access to the Flipnote software, so this Krita package is the most accurate parallel to the brushes on Flipnote that are readily available, to my knowledge.

## Concerns
I don't have much technical knowledge of what goes into making custom brushes, so I'm not sure how viable these dithering brushes are to make or use. I have tried to make custom raster brushes in the past but to no avail. 

## Enticement
On my path to try and replicate the style, I came across a like minded individual who also ran into this problem, CS Ghost Animation (Youtuber). Not that it adds much, but I'm sure he would appreciate the availability of these brushes as well, since he has to copy and paste in order to obtain the dithering effect seen in Flipnote.

**Comment by RodneyBaker** [comment]
I downloaded the brush pack and am looking into it.
In the meantime, have you contacted the author of the brushes?
They would be the first best place to get advice on porting the brushes over to Opentoonz.
They might even know a little about how to get that done.

Disclaimer:  I am not at all familiar with Flipnote and have very basic understanding of Krita.

**Comment by GoodbyeGhost** [comment]
True, I was looking at contacting them, I'll see if they can help port it to Opentoonz. It's not exactly clear how they implemented them into Krita, so I'm sure they'll have some good insight

**Comment by RodneyBaker** [comment]
Any information they can provide will help.
If we are going to port the brushes it'd be best to do it in a way that the brushes meet their full potential.

**Comment by GoodbyeGhost** [comment]
Still waiting on a response form the author, but in the meantime, I've been looking at the brushes more. I think there are only three things to worry about when trying to port the pack to Opentoonz. 1: The dither brushes need to feel like they're revealing a pattern, see images 1-3. 2: When brush size increases or decreases, it needs to adjust how much it reveals rather than change the pattern, gif below. 3: The brush shouldn't overwrite what's underneath it, only add to it, nother gif below.

If we can accomplish those things than we'd need to worry about what texture to feed Opentoonz to gain the same visual effect. This part I'm not so sure on. Depending on how Opentoonz works, we can either take Sabs' route with high resolution images, or just use the bare minimum pixels needed for the pattern. If I had to guess from what I've seen, I think the high resolution canvases Sabs used was for when brush sizes changed, but I don't know how that would work in Opentoonz. Gonna see if I can figure out the reason for those large textures in the Flipnot Brush Kit in the meantime.

![1](https://user-images.githubusercontent.com/77942602/150703006-095aa8bb-46d5-4076-9e3c-2750d34c0ea9.png)
![2](https://user-images.githubusercontent.com/77942602/150703009-839ef9df-2686-482d-b1ed-113b90ea5bf0.png)
![3](https://user-images.githubusercontent.com/77942602/150703011-13570f79-e9cf-4b63-a93b-a17a42b39c11.png)
![size](https://user-images.githubusercontent.com/77942602/150703019-cca951a4-787e-4361-812e-a1b0c98f11c8.gif)
![overlap](https://user-images.githubusercontent.com/77942602/150703023-03d4747d-3c52-4130-bd92-f31e1b3fca7e.gif)

**Comment by RodneyBaker** [comment]
Interesting.
I'd say that will require a new brush engine of some type.

In the meantime we can do similar things with other approaches due to layering and such.
For instance, we can set any raster level to have white as transparent.  So drawing with white will reveal any pattern or image layered behind it.  How to get it to shift to different layer/pattern is a  bit more complicated but can also be done to some extent although not in the context of a brush.  

In short, I don't see your brushes as demonstrated being implemented any time soon but in such cases I'm always happy to be wrong.

**Comment by Bracket-H** [comment]
There's always the option of drawing them elsewhere and importing the final drawings into OT.

**Comment by RodneyBaker** [comment]
That does seem to be the primary option at this time @Bracket-H 
One could even do line work in Opentoonz and then do the Flipnote style patterning in Krita.
That does seem the best way forward.

**Comment by artisteacher** [comment]
As an alternative, you could make an image texture and load it as a texture style. If you change the position to "fixed", it'll behave just as you want it to on toonz raster levels.  Though you'll only see the texture when preview is turned on.

**Comment by manongjohn** [comment]
I would suggest looking into making them as MyPaint brushes as there is more control over how the style behaves. There may be a MyPaint brushes in 1 or more sets that are similar and can be tweeked to mimic the Flipnote style.

**Comment by GoodbyeGhost** [comment]
Been very interested in implementing the feature fully instead of taking another one of these routes. I've tried making custom MyPaint brushes and using textures but the results don't satisfy me. Not sure if MyPaint is capable of doing this, else I probably would go that route. I'm thinking I can find a way to reverse some of the texture code into doing what I want, though the codebase isn't exactly a breeze to sift through, so yea, this is gonna take a while to get right.

**Comment by GoodbyeGhost** [comment]
True, I've used other software in tandem with Opentoonz in the past, but for animation I always come back to it. Pixel art software I especially can't stand in terms of animation, and only having Krita as an alternative just doesn't sit well with me. Obviously I don't know how difficult implementing something like this is just yet, so I'm more than likely being more stubborn than I need to be. The way I see it, there may be many ways to go about implementing something like this that don't seem obvious from the start, but I won't know that until I fully explore the idea. 

By myself I'm sure this'll take long, so that's why I brought the idea here. I'm hoping that someone with more experience comes along and wants to full send on this idea as well. Then again, it's hard to describe why I want this in the first place without seeing how niche of an idea it is. I very much like the style of flipnotes and very much like Opentoonz, so I want to bring dithering into the picture, as it's the only thing stopping me from using Opentoonz throughout the whole process. I'll have to think about how I can broaden the issue into something more helpful to the software than the niche.

**Comment by Bracket-H** [comment]
Here, have a quick video on it, sorry for the horrendous drawing and all that, this is not my actual work computer (monitor is much smaller) etc, but maybe you'll see the point and potential benefit, including the one at the end where I show off that by using texture, you can actually animate the texture as well (and you can keyframe color too)

And once you have set it up, even with more indices, you can just save the whole chain of texture effects as a macro fx, and then save that one as a preset, and then you'll have a ready to go set for any future project that will adapt to your indices, colors etc (according to the params you saved the preset as + its input slots)

P.S.: Be sure to create a texture that is as big as your output, I didn't do that cause quick demo video...
And again, imagine it with actual artwork, and actual textures scaled to your liking. (I did a quick xpaint dump of its patterns which are very small even on 1080p )
https://files.catbox.moe/89u4w3.mp4

**Comment by GoodbyeGhost** [comment]
I like this method, but there are some things that *ideally* I would want to replicate with the dithering that doesn't seem possible like this. 

First would be a color mixing effect, where two separate dither patterns with two separate colors are drawn atop each other combining into another color. One of the ways to mimic this with the Krita brush pack is to use the Shogi and Biscuit brushes on top of each other with separate colors. Second is distorting the drawn pixels from the texture, like when transforming the image or applying filters. I plan on also making custom fx to mimic some filters found on flipnote, which should work even with this method, though transforming the pixels would still keep the patterns in line with the source texture. There's an example youtube vid below of these techniques (keep note that purple isn't a color normally available in Flipnote without using the first method).

Another problem is making and keeping track of styles or layers for each color of each pattern I use. I'm sure I could handle that somehow, though I'd have to use this method in an animation to see how I feel about it first. If we're speaking long term, I was wondering if doing something like what Aseprite does with selection brushes could be plausible in OT?

(Also sorry if poorly formatted, had to rush this comment)

(pattern stays in place) 
https://files.catbox.moe/ly11v5.mp4
(example vid - WARNING, probs flashing lights and such)
https://www.youtube.com/watch?v=31JDgKylk24

**Comment by Bracket-H** [comment]
Ah, the pattern rotation is a weakness of this metod for sure. Can flipnote do that, though? (I have never used it)
Technically you could 'bake' the output of the effects into new levels and use them as source you can then work out more naturally (including rotation) but even I have to admit that that sounds a bit like a pain.

The color blending thing, however, should work rather well, especially if you make a toonz raster dither texture to use as the effect texture.
Why? Because then you'd get every single benefit of that format, including keyframing the color of the dithers itself, and of course plugging that into other fx nodes as well.

Again, I agree that that is not as easy to use as a direct brush you paint dithers with...
And the native Toonz raster brushes don't have textures...because they're intended to be (aliased), 'meta data' carrying images that are easily processed further.
(Hence fancy outlines and the like being effect nodes).
But, this also gives you power over the dithers that flipnote probably can't do. 
(and probably don't even want, but you could for example, attach a tiny little lens flare to every dither point, or make god rays shine through the mask, blur it, pulse it, whatever)

Anyway, long story short, you got me on the rotating patterns, that's possible too (probably with pinned texture, can't test right now) but again, not as straightforward as painting patterns directly...which the native raster probably won't do any time soon.

Technically the format can deal with it, but if it were easy, I think it would be in already and not outsourced to a different raster format through mypaint.

**Comment by GoodbyeGhost** [comment]
To clarify on the pattern rotation thing, Flipnote only allows three layers so things tend to get baked down in the animation process. Lines probably end up with the dithering in the end, if not used for background or overlay. It's difficult to find documentation on the process most of these creators take, or the exact tools Flipnote has, so I'm making guesses of how it all works from the Flipnote animations, some tips from the fandom wiki, Clipnote Studio, and occasionally some old tutorial vid I find on Youtube.

Considering this, I'm not to apposed to the idea of baking the effects. I'll have to do some playing around with the FX Schematic, see if I can come up with a process or maybe other methods to explore. 

On another note, sounds like that issue you mentioned is the one we should be solving right now... again, not acquainted with the code, but that sounds like the dreaded "pea under the mattress" situation. I wonder if the solution to that problem is tucked beneath 10 or 100 ft of code.

**Comment by RodneyBaker** [comment]
Closing this but will tag with 'research'.
The usefulness of the pattern brushes should be obvious... how best to proceed is the question.

**Comment by OstojaOfficial** [comment]
Any updates on this?

**Comment by GoodbyeGhost** [comment]
Nothing to my knowledge. I haven't been active on this issue for a while. Was looking into porting effects that mimic some of the tecchniques used in Flipnote animations though. While I don't think I can replicate the brushes just yet, effects could be a step in the right direction. One example is a Bayer matrix filter, like the camera uses in Flipnote, which I imagine would be the most widely used effect besides grow or shrink. Another issue that arose is the fx schematic itself. I think there are some quality of life improvements that can be made to make it easier to reproduce results, and possibly reduce friction between animation and post-processing phases when making complex animations. I'll have to get my thoughts organized on this, because I've been wanting to contribute for quite a while now.

**Comment by andeon** [comment]
I'm starting to work on this, not a brush but a screentone set pack. I need to verify that this workflow doesn't overflow OpenToonz, so not sure if I'm going to try to push to the project or just open a repository, as it seems like all of this is something very specific.

Initial test:

https://github.com/opentoonz/opentoonz/assets/11843239/b0cd85bc-5cda-4a94-9efa-7f8e0ab8402f

If you guys can list the most desirable effects for me so I can create/mimic and test the workflow to make this start to happen.

![test](https://github.com/opentoonz/opentoonz/assets/11843239/c46d57de-fbc7-4fe9-a4c6-e73958d1b978)

**Comment by deathchimera** [comment]
@andeon Just saw this while browsing PRs, good stuff. If you're looking for Flipnote replication, then there are three major things preventing authentic simulation of it.

1. **Overlapping of textures/dithering** (might only be possible through seperate layers currently, not sure. Used for combining colors and base dithering to make new dither patterns). 
2. **Texture baking** (Flipnotes often bake the dithering into the drawing layer for movements, so the dithering gets distorted with the drawings). 
3. **Color manipulation of textures** (Your textures play well with colors, but I had trouble with this in the past).

Looking at your PR, you seem to have addressed as much of the problems listed as you can using Textures, and I'm not sure how much further you can go with a screentone pack if you want to replicate Flipnote. 

Not sure if it'll help, but I did make a texture pack in the past while trying to replicate Flipnote. I couldn't get the colors to change the texture color, so I thought changing the color of a texture wasn't possible (or something like that, it's been a while). Looking at your textures, I realized that my textures weren't working because the pixels were pure black. 

Here is my version, using the least ammount of space I could. I think these are all the brushes available in Flipnote, or at least the Krita Flipnot pack I use. Feel free to include this in your PR if you'd like, not sure how to go about that.

[FlipnotePatternsOT.zip](https://github.com/opentoonz/opentoonz/files/11533532/FlipnotePatternsOT.zip)

**Comment by andeon** [comment]
The pack covers all these effects. If you try to change the rotation option as well, you'll get new effects, but it might be a bit blurry.

I'm checking if I can also create a shader FX to cover another way to do this.

If we could enable the texture for use on the Raster level type "the blue one", we could achieve this effect in realtime with Overlapping/Dithering and so on. Many programs support this feature, perhaps nobody had thought about this possibility.
